kf/kf_unix.go

```diff
@@ -301,58 +365,143 @@ func (b *BPF) LoadTCAttachProgram(ifaceName, direction string) error {
 
 // UnloadTCProgram - Remove TC filters
 func (b *BPF) UnloadTCProgram(ifaceName, direction string) error {
-
 	iface, err := net.InterfaceByName(ifaceName)
 	if err != nil {
 		log.Error().Err(err).Msgf("UnloadTCProgram - look up network iface %q", ifaceName)
 		return err
 	}
 
-	bpfRootProg := b.ProgMapCollection.Programs[b.Program.EntryFunctionName]
-
-	var parent uint32
-	if direction == models.IngressType {
-		parent = tc.HandleMinIngress
-	} else if direction == models.EgressType {
-		parent = tc.HandleMinEgress
+	tcgo, err := tc.Open(&tc.Config{})
+	if err != nil {
+		log.Error().Err(err).Msgf("UnloadTCProgram - Unable to tc.Open(&tc.Config{}):  %q", ifaceName)
+		return err
 	}
 
-	tcfilts, err := b.TCFilter.Get(&tc.Msg{
-		Family:  unix.AF_UNSPEC,
-		Ifindex: uint32(iface.Index),
-		Handle:  0x0,
-		Parent:  core.BuildHandle(tc.HandleRoot, parent),
-	})
-
+	clsactFound := false
+	htbFound := false
+	ingressFound := false
+	var htbHandle uint32
+	var ingressHandle uint32
+	var parentHandle uint32
+	// get all the qdiscs from all interfaces
+	qdiscs, err := tcgo.Qdisc().Get()
 	if err != nil {
-		log.Warn().Msgf("Could not get filters for interface \"%s\" direction %s ", ifaceName, direction)
-		return fmt.Errorf("could not get filters for interface %s : %v", ifaceName, err)
+		return fmt.Errorf("could not get qdiscs for interface %s : %v", ifaceName, err)
+	}
+	for _, qdisc := range qdiscs {
+		iface, err := net.InterfaceByIndex(int(qdisc.Ifindex))
+		if err != nil {
+			return fmt.Errorf("could not get interface %s from id %d: %v", ifaceName, qdisc.Ifindex, err)
+		}
+		if iface.Name == ifaceName && qdisc.Kind == "clsact" {
+			clsactFound = true
+		}
+		if iface.Name == ifaceName && qdisc.Kind == "htb" {
+			htbFound = true
+			htbHandle = qdisc.Msg.Handle
+		}
+		if iface.Name == ifaceName && qdisc.Kind == "ingress" {
+			ingressFound = true
+			ingressHandle = qdisc.Msg.Handle
+		}
```

Comment:
can we convert this to switch statement as above.

Reply:
Fixed in the newer commit